### PR TITLE
SE-610: Adding support for bond instrument in load_from_data_frame

### PR DESCRIPTION
**Pull request by cscialino**
SE-610: Adding support complex instruments in load_from_data_frame

# Pull Request Checklist

- [ x ] Read the [contributing guidelines](https://github.com/finbourne/lusid-python-tools/blob/master/docs/CONTRIBUTING.md)
- [ x ] Tests pass

# Description of the PR

- Added support for Bond Instrument definition to ```load_from_data_frame```
- Tests passed

### CODE REVIEW

File: tests/unit/cocoon/test_cocoon_complex_instrument.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+        combined_successes = {
+            correlation_id: instrument
+            for response in responses["instruments"]["success"]
+            for correlation_id, instrument in response.values.items()
+        }
+
+        self.assertTrue(
+            expr=all(
+                sorted(combined_successes[key].properties, key=lambda p: p.key)
```
Comment (MikeMcGarry) [inline tests/unit/cocoon/test_cocoon_complex_instrument.py:196]: Does this test illustrate that your change is working, it looks to only check that all the property keys are the same between the test and the expected outcome?

File: tests/unit/cocoon/test_cocoon_complex_instrument.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+                {"Isin": "ISIN", "ClientInternal": "Client Internal"},
+                [],
+                "TestPropertiesScope1",
+            ],
+            [
+                "A standard successful load of a complex instrument using a full bond definition",
+                "TestScope1",
+                "data/global-fund-fixed-income-master.csv",
+                {
+                    "name": "Name",
```
Comment (MikeMcGarry) [inline tests/unit/cocoon/test_cocoon_complex_instrument.py:103]: I don't see where you would ever have a dictionary which has `instrument_type`  as a key based on this mapping here?

File: lusidtools/cocoon/utilities.py [hunk 3]
```diff
@@ -816,9 +817,28 @@ def handle_nested_default_and_column_mapping(
 
         # If the value of the mapping is a dictionary
         if isinstance(value, dict):
+            nested_key = key
+
+            # If an 'instrument_type' is specified within mapping, expand nested definition mapping
+            if "instrument_type" in mapping.keys():
```
Comment (MikeMcGarry) [inline lusidtools/cocoon/utilities.py:823]: Can we make this agnostic to instruments?

File: lusidtools/cocoon/utilities.py [hunk 4]
```diff
@@ -816,9 +817,28 @@ def handle_nested_default_and_column_mapping(
...
+                    models.LusidInstrument.discriminator_value_class_map
+                )
+                if mapping_object not in valid_instrument_types:
+                    raise TypeError(
+                        f"""The specified instrument type {mapping_object} is invalid.
+                        Please set instrument type to one of: {valid_instrument_types.keys()}"""
+                    )
+                else:
+                    for key, value in mapping[nested_key].items():
+                        mapping_updated[key] = value
```
Comment (MikeMcGarry) [inline lusidtools/cocoon/utilities.py:838]: Even if the mapping updated how would this get used later on? There is no field called `dom_ccy` as per your test the mapping would need to be `definition.dom_ccy`.

File: lusidtools/cocoon/utilities.py [hunk 5]
```diff
@@ -260,27 +260,40 @@ def set_attributes_recursive(
...
+            # Handle identifiers provided within instrument definition (e.g. 'Bond', 'Future', etc.)
+            if (key, attribute_type) == ("identifiers", "dict(str, str)"):
+                obj_init_values[key] = {
+                    str_key: row[str_value["value"]]
+                    for str_key, str_value in mapping[key].items()
+                }
+                continue
+            else:
+                obj_init_values[key] = additional_attributes[key]
+                continue
```
Comment (MikeMcGarry) [inline lusidtools/cocoon/utilities.py:277]: You will continue here no matter what, so you can put continue just once outside the if/else on line 277 and remove it on line 274.

File: lusidtools/cocoon/utilities.py [hunk 6]
```diff
@@ -260,27 +260,40 @@ def set_attributes_recursive(
...
 
+        # Get the attribute type
+        attribute_type = obj_attr[key]
+
         # If it is an additional attribute, populate it with the provided values and move to the next attribute
         if key in list(additional_attributes.keys()):
-            obj_init_values[key] = additional_attributes[key]
-            continue
+            # Handle identifiers provided within instrument definition (e.g. 'Bond', 'Future', etc.)
+            if (key, attribute_type) == ("identifiers", "dict(str, str)"):
```
Comment (MikeMcGarry) [inline lusidtools/cocoon/utilities.py:269]: Ahhh very nice, great pickup that we are considering `identifiers` to be an exception to how we process most attributes,  clearly with these `LusidInstruments` that exception is broken. This seems to be a sensible way of handling the exception to the exception.

File: lusidtools/cocoon/utilities.py [hunk 7]
```diff
@@ -260,27 +260,40 @@ def set_attributes_recursive(
...
     for key in list(populate_attributes):
 
+        # Get the attribute type
+        attribute_type = obj_attr[key]
+
         # If it is an additional attribute, populate it with the provided values and move to the next attribute
         if key in list(additional_attributes.keys()):
-            obj_init_values[key] = additional_attributes[key]
-            continue
+            # Handle identifiers provided within instrument definition (e.g. 'Bond', 'Future', etc.)
```
Comment (MikeMcGarry) [inline lusidtools/cocoon/utilities.py:268]: Love this comment, very useful thank you!

File: lusidtools/cocoon/utilities.py [hunk 8]
```diff
@@ -260,27 +260,40 @@ def set_attributes_recursive(
...
+            if (key, attribute_type) == ("identifiers", "dict(str, str)"):
+                obj_init_values[key] = {
+                    str_key: row[str_value["value"]]
```
Comment (MikeMcGarry) [inline lusidtools/cocoon/utilities.py:271]: Nice dictionary comprehension!

Comment (MikeMcGarry) [inline lusidtools/cocoon/utilities.py:271]: Why do we take `str_value["value"]`? My understanding is that `str_value` is a string, wouldn't calling `str_value["value"]` result in an Exception being thrown?

File: lusidtools/cocoon/utilities.py [hunk 9]
```diff
@@ -260,27 +260,40 @@ def set_attributes_recursive(
...
+                    for str_key, str_value in mapping[key].items()
```
Comment (MikeMcGarry) [inline lusidtools/cocoon/utilities.py:272]: Do we need to add `if not pd.isna(row[mapping[key]])` to ensure that we don't attempt to get a value from the DataFrame when it is `NaN` or `None`?

File: lusidtools/cocoon/utilities.py [hunk 10]
```diff
@@ -260,27 +260,40 @@ def set_attributes_recursive(
...
             # If this exists in the mapping with a value and there is a value in the row for it
             if mapping[key] is not None and not pd.isna(row[mapping[key]]):
                 # Converts to a date if it is a date field
                 if "date" in key or "created" in key or "effective_at" in key:
                     obj_init_values[key] = str(DateOrCutLabel(row[mapping[key]]))
+                # Converts to a list element if it is a list field
+                elif "list" in attribute_type and not isinstance(
+                    row[mapping[key]], list
+                ):
+                    obj_init_values[key] = list(row[mapping[key]].split())
```
Comment (MikeMcGarry) [inline lusidtools/cocoon/utilities.py:296]: Does `obj_init_values[key] = [row[mapping[key]]]` work?  That is what we use on line 318 below and it is simpler than splitting.

File: lusidtools/cocoon/utilities.py [hunk 11]
```diff
@@ -314,7 +327,24 @@ def set_attributes_recursive(
...
     # Create an instance of and populate the model object
-    return model_object(**obj_init_values)
+    instance = model_object(**obj_init_values)
+
+    # Support for polymorphism, we can identify these `abstract` classes by the existence of the below
+    if getattr(instance, "discriminator"):
+        discriminator = getattr(instance, getattr(instance, "discriminator"))
+        try:
+            actual_class = model_object.discriminator_value_class_map[discriminator]
+        except TypeError:
```
Comment (MikeMcGarry) [inline lusidtools/cocoon/utilities.py:337]: This will likely be a `KeyError` rather than a `TypeError`.

File: lusidtools/cocoon/utilities.py [hunk 12]
```diff
@@ -314,7 +327,24 @@ def set_attributes_recursive(
...
+            print(
+                f"Invalid discriminator object {discriminator} was passed. "
+                f"Please ensure is one of: {model_object.discriminator_value_class_map}"
+            )
```
Comment (MikeMcGarry) [inline lusidtools/cocoon/utilities.py:341]: Great checking, can we please use `logging.error` instead of `print`?

File: tests/unit/cocoon/test_cocoon_complex_instrument.py [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+from parameterized import parameterized
+import lusid
+import lusid.models as models
+from lusid.utilities import ApiClientFactory
+from lusidtools import logger
+from datetime import datetime
+import pytz
+
+
+class ComplexInstrumentTests(unittest.TestCase):
```
Comment (MikeMcGarry) [inline tests/unit/cocoon/test_cocoon_complex_instrument.py:15]: Great test, can you please add another test to test the failure case where the discriminator isn't valid?

**Review by MikeMcGarry — CHANGES_REQUESTED** [review]
Looking great, fantastic job!

A couple of places I think we can make some slight optimisations which I have made comments for.

On line 271 I need to understand how ` row[str_value["value"]]` works.

And finally can we please add a test for the failure case in which you provide an invalid instrument type e.g. "UnsupportedInstrument" to ensure we understand what the behaviour will be and assert that the error raised is as we expect.

File: lusidtools/cocoon/utilities.py (hunk 12 above)
Comment (cscialino) [reply lusidtools/cocoon/utilities.py:341]: Certainly!

File: lusidtools/cocoon/utilities.py (hunk 9 above)
Comment (cscialino) [reply lusidtools/cocoon/utilities.py:272]: Yes you are right. In fact something to add a test for, and I think we'll want to ensure it opens an empty dict() when it's NaN. Otherwise, one is required to map the identifier a second time, which is a mistake that people will surely make given the variable is set as nullable.

Comment (MikeMcGarry) [reply lusidtools/cocoon/utilities.py:272]: Yes, great point about a dictionary if all are NaN! I wonder if that would be the behaviour anyway if all are NaN or not? Like you say best way to see is via a test!
